crates/ra_hir_ty/src/_match.rs

```diff
@@ -530,6 +532,16 @@ pub(crate) fn is_useful(
     matrix: &Matrix,
     v: &PatStack,
 ) -> MatchCheckResult<Usefulness> {
+    // Handle the special case of enums with no variants. In that case, no match
+    // arm is useful.
+    if let Ty::Apply(ApplicationTy { ctor: TypeCtor::Adt(AdtId::EnumId(enum_id)), .. }) =
+        cx.infer[cx.match_expr].deref()
+    {
+        if cx.db.enum_data(*enum_id).variants.is_empty() {
+            return Ok(Usefulness::NotUseful);
+        }
+    }
```

Comment:
I'd expect this special case to become unnecessary when we properly determine constructors based on the type, but for now it's fine

Reply:
Agreed :+1: 

The issue with an enum with no variants is our current method relies on the existence of at least one match arm in order to determine the constructor, which is what makes this a special case. If we always used the match expr to determine the construtor we wouldn't have that problem.